src/bp/core/services/cms.ts

```diff
@@ -578,15 +578,49 @@ export class CMSService implements IDisposeOnExit {
 
   // This methods finds the translated property and returns the original properties
   private getOriginalProps(formData: object, contentType: ContentType, lang: string, defaultLang?: string) {
-    const originalProps = Object.keys(_.get(contentType, 'jsonSchema.properties'))
+    const schema = <FormDefinition>contentType.newSchema
+    const fields = [...schema.fields, ...schema.advancedSettings]
+    const result = _.cloneDeep(formData)
 
-    if (originalProps) {
-      return originalProps.reduce((result, key) => {
-        result[key] = formData[key + '$' + lang] || (defaultLang && formData[key + '$' + defaultLang])
-        return result
-      }, {})
-    } else {
-      return formData
+    this.applyTranslations(result, fields, lang, defaultLang)
+
+    return result
+  }
+
+  private applyTranslations(current: any, fields: FormField[], lang: string, defaultLang?: string) {
+    for (const field of fields) {
+      if (!current[field.key]) {
+        continue
+      }
+
+      if (field.translated) {
+        current[field.key] = current[field.key][lang]
+      }
+
+      if (field.fields) {
+        let fields = field.fields
+
+        for (const subfield of field.fields) {
+          if (subfield.options) {
+            for (const option of subfield.options) {
+              if (option.related) {
+                if (fields === field.fields) {
+                  fields = [...field.fields]
+                }
+                fields.push(option.related)
+              }
+            }
+          }
+        }
```

Comment:
This works but it's a bit dirty. I think it would be beneficial to preprocess a more efficient object for each content type to make translations faster and also simplify the code.

Reply:
The translate logic can easily be schema agnostic. What happens if we add unknown properties ? They will not be translated. As I told you, i've already updated the logic to handle the new format in another PR because I needed it, it just need to be added here.

https://github.com/botpress/botpress/blob/prompts/src/bp/core/misc/templating.ts#L8

What needs to be done is to update the migration v12.11.0 (migrate say node) so it uses the new schema